{"metadata":{"kernelspec":{"display_name":"R","language":"R","name":"ir"},"language_info":{"mimetype":"text/x-r-source","name":"R","pygments_lexer":"r","version":"3.4.2","file_extension":".r","codemirror_mode":"r"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# A visual exploration of the size and GC content of sequenced genomes avaliable on [NCBI](https://www.ncbi.nlm.nih.gov/)\n\n   #Genome size is commonly reported in megabase pairs (megabase = 1,000,000 base pairs of DNA). The size of genomes is extremly variable across species and influenced by a large number of factors (if you're curious the human genome is ~3,000 Mb (megabase pairs), or 3 billion base pairs in size). The NCBI sequenced genome dataset includes genomes from three domains of organisms: eukaryotes (multi cellular organisms, like you, ducks and ferms), porkaryote (single cell, bacteria) and virus genomes (not a Linnean doman... [more a weird grey area](https://www.scientificamerican.com/article/are-viruses-alive-2004/). Here I plot the size of the genomes in these three domains to compare them, and also look at the GC content of the genomes across these domains. %GC content is the percentage of DNA base pairs composed of G and C (guanine and cytosine) GC-content is indicative of many protein coding genes, and the %GC content is indicative of a gene-rich genome. GC ratios within a genome are know to be extremly variable, with low %GC indicating a large amount of non-coding DNA in the genome. \n\n#These data are derived from [NCBI's repository of sequenced genomes](https://www.ncbi.nlm.nih.gov/home/genomes/). Through the links provided in the Kaggle dataset, one can access the entire sequence to any of these genomes as they are pubically avaliable (for free! thanks science!)","metadata":{"_uuid":"847638d7b6aaeaa56c7032d212ad9fb98f6dc4c6","_cell_guid":"9b230c27-ba5b-45a8-8f26-db2955642650","execution":{"iopub.status.busy":"2022-05-13T00:09:40.673213Z","iopub.execute_input":"2022-05-13T00:09:40.674272Z","iopub.status.idle":"2022-05-13T00:09:40.684866Z"}}},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Loading the data, assessment and cleanup\n","metadata":{"_uuid":"35cfd25bf3e26110296bba7f636efca0e6e49e57","_cell_guid":"1e8cee89-39c1-450c-8ad4-3f33c755322d"}},{"cell_type":"code","source":"library('tidyverse')\nlibrary('ggthemes')","metadata":{"_execution_state":"idle","_uuid":"fee0228231cf3d013915cfa42f302d6514f51e69","_cell_guid":"73dd573b-e12a-421e-aaf4-ed8b3398d4a8","execution":{"iopub.status.busy":"2022-05-19T03:37:02.781098Z","iopub.execute_input":"2022-05-19T03:37:02.785229Z","iopub.status.idle":"2022-05-19T03:37:02.797607Z"},"trusted":true},"execution_count":155,"outputs":[]},{"cell_type":"code","source":"raw_eukaryote = read_csv('../input/eukaryotes.csv')\nhead(raw_eukaryote)","metadata":{"scrolled":true,"_uuid":"a6bfd45fe6fb882ad26c564245d4f4a7c1a2a788","_cell_guid":"d781bafa-b848-46b0-9c6f-8af10db645a0","execution":{"iopub.status.busy":"2022-05-19T03:37:02.801132Z","iopub.execute_input":"2022-05-19T03:37:02.803633Z","iopub.status.idle":"2022-05-19T03:37:02.933757Z"},"trusted":true},"execution_count":156,"outputs":[]},{"cell_type":"code","source":"# just the chromosome and scaffold level assembies, the contigs removed\neukaryote = raw_eukaryote[raw_eukaryote$Level == \"Chromosome\" || raw_eukaryote$Level == \"Scaffold\",]\nhead(eukaryote)","metadata":{"scrolled":true,"_uuid":"4fd42e5d21225b141111471c5a2346032e28f671","_cell_guid":"a527a8b0-86d5-4481-8b8a-0e6716f65d6d","execution":{"iopub.status.busy":"2022-05-19T03:37:02.936290Z","iopub.execute_input":"2022-05-19T03:37:02.938311Z","iopub.status.idle":"2022-05-19T03:37:02.987998Z"},"trusted":true},"execution_count":157,"outputs":[]},{"cell_type":"markdown","source":"### read in the eukaryote data","metadata":{"_uuid":"1ced7389eaf5a37e26ba58e0b99f4c760ea189d6","_cell_guid":"b3a8a3aa-9bbb-4cef-be9e-145006d1915d"}},{"cell_type":"code","source":"eukaryote = eukaryote[eukaryote$\"GC%\" > 5,]\n#Prototheca stagnorum has 71% GC content which is pretty neat\neukaryote[eukaryote$\"GC%\" > 70,]","metadata":{"_uuid":"3cc2aaa64f0d83df0c6f7fc9b7bd521dd7c0ecd9","_cell_guid":"270e4339-5f7a-4837-9d8c-98436c5c3d7e","execution":{"iopub.status.busy":"2022-05-19T03:37:02.990648Z","iopub.execute_input":"2022-05-19T03:37:02.993667Z","iopub.status.idle":"2022-05-19T03:37:03.000566Z"},"trusted":true},"execution_count":158,"outputs":[]},{"cell_type":"code","source":"#looking at the data there appears to be some missing data.\n# GC of zero seen, looks like one with with a very low number,\n# also an outlier with >70% gc, I should look at what this one is \nqplot(eukaryote$\"GC%\") +\n    labs(title = \"Histogram of Eukaryote genome %GC content\", x= \"%GC content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"_uuid":"d8b84411dbfec9b643975e2319bed2389c29e224","_cell_guid":"3864b316-60e2-42da-894e-61a2780b1584","execution":{"iopub.status.busy":"2022-05-19T03:37:03.003276Z","iopub.execute_input":"2022-05-19T03:37:03.006921Z","iopub.status.idle":"2022-05-19T03:37:03.289894Z"},"trusted":true},"execution_count":159,"outputs":[]},{"cell_type":"markdown","source":"### read in the prokaryote data","metadata":{"_uuid":"5a1e4c50b4ffbf5fc4718d5dcaf600ac0362b85a","_cell_guid":"2aba6bfd-fdb6-462a-9870-cb5fa3d80b04"}},{"cell_type":"code","source":"prokaryote = read_csv('../input/prokaryotes.csv')\nprokaryote = prokaryote[prokaryote$\"GC%\" > 5,]","metadata":{"_uuid":"e6bf165689f176a4f20298c2bc2a1857da574c15","_cell_guid":"4ed5c5c9-58a5-4ffd-81be-89082c4f3f1c","execution":{"iopub.status.busy":"2022-05-19T03:37:03.292555Z","iopub.execute_input":"2022-05-19T03:37:03.295169Z","iopub.status.idle":"2022-05-19T03:37:03.398895Z"},"trusted":true},"execution_count":160,"outputs":[]},{"cell_type":"code","source":"#check to make sure things aren't valuable\nlow_prokaryote = prokaryote[prokaryote$\"GC%\" < 20,]\nhead(low_prokaryote)","metadata":{"_uuid":"0d4e08e554ee613287c2265c3b56efadbca713c0","_cell_guid":"2eba28e4-df90-4a8b-a45d-f6dfd015cb12","execution":{"iopub.status.busy":"2022-05-19T03:37:03.401834Z","iopub.execute_input":"2022-05-19T03:37:03.404373Z","iopub.status.idle":"2022-05-19T03:37:03.451762Z"},"trusted":true},"execution_count":161,"outputs":[]},{"cell_type":"code","source":"#wider distribution for prokaryotes\n# higher mean and median\nqplot(prokaryote$\"GC%\") +\n    labs(title = \"Histogram of Prokaryote genome %GC content\", x= \"%GC content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"scrolled":true,"_uuid":"51a02806f5eb302ff0fcb665b7e5f7cca485818c","_cell_guid":"c1f7a396-4108-4fe0-855d-c440680b05e3","execution":{"iopub.status.busy":"2022-05-19T03:37:03.454195Z","iopub.execute_input":"2022-05-19T03:37:03.456716Z","iopub.status.idle":"2022-05-19T03:37:03.753202Z"},"trusted":true},"execution_count":162,"outputs":[]},{"cell_type":"markdown","source":"### read in the virus data","metadata":{"_uuid":"9d28be227722bf3805116622a3ad7900a7e98bcc","_cell_guid":"5886bde7-b385-4f22-ac1e-a23698b0afba"}},{"cell_type":"code","source":"#one 0 in the viruses... also a gc as high as 80+ investigate further\nviruses = read_csv('../input/viruses.csv')\nviruses = viruses[viruses$\"GC%\" > 5,]\nhead(viruses)","metadata":{"_uuid":"a85f4af09f7ad423f8c4df70029149c5a353168c","_cell_guid":"c53fde9d-e318-45b1-be69-3b5b4fa6ad42","execution":{"iopub.status.busy":"2022-05-19T03:37:03.755701Z","iopub.execute_input":"2022-05-19T03:37:03.758162Z","iopub.status.idle":"2022-05-19T03:37:03.867924Z"},"trusted":true},"execution_count":163,"outputs":[]},{"cell_type":"code","source":"qplot(viruses$\"GC%\")+\n    labs(title = \"Histogram of Virus genome %GC content\", x= \"%GC content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"_uuid":"a4f5e8a265b5d0bfd32db4a7d48e7090c1393dfa","_cell_guid":"1455fc63-bece-4a17-abc3-1f9625067504","execution":{"iopub.status.busy":"2022-05-19T03:37:03.870478Z","iopub.execute_input":"2022-05-19T03:37:03.873084Z","iopub.status.idle":"2022-05-19T03:37:04.188613Z"},"trusted":true},"execution_count":164,"outputs":[]},{"cell_type":"markdown","source":"## %GC content - comparison of means and medians\n\nAs a first step we can assess the means and medians for the different organism types.","metadata":{"_uuid":"84ad6c4779e2b152ef5a99e001980ba6cbce816f","_cell_guid":"af565c22-10aa-4a2d-9b06-8749ffd56a53"}},{"cell_type":"code","source":"mean(eukaryote$\"GC%\")","metadata":{"_uuid":"78d41c0648efbdaa834920a239eaae09264a9de0","_cell_guid":"00d6e0ae-794d-4c48-b6ba-2905cbe5facb","execution":{"iopub.status.busy":"2022-05-19T03:37:04.199775Z","iopub.execute_input":"2022-05-19T03:37:04.202451Z","iopub.status.idle":"2022-05-19T03:37:04.217583Z"},"trusted":true},"execution_count":165,"outputs":[]},{"cell_type":"code","source":"mean(prokaryote$\"GC%\")","metadata":{"_uuid":"a596b7a311ccaaaa7bb23843fe93dcfea88e34e5","_cell_guid":"29847cea-ab44-4673-825a-756fe27362ad","execution":{"iopub.status.busy":"2022-05-19T03:37:04.220726Z","iopub.execute_input":"2022-05-19T03:37:04.223202Z","iopub.status.idle":"2022-05-19T03:37:04.236461Z"},"trusted":true},"execution_count":166,"outputs":[]},{"cell_type":"code","source":"#mean of this GC in between the two others\nmean(viruses$\"GC%\")","metadata":{"_uuid":"8cfe7f480b61125781f799f0652be6ee51e21fb0","_cell_guid":"c14bb67f-db6d-4f4d-a91e-bf3a4a2e2848","execution":{"iopub.status.busy":"2022-05-19T03:37:04.238698Z","iopub.execute_input":"2022-05-19T03:37:04.242020Z","iopub.status.idle":"2022-05-19T03:37:04.256189Z"},"trusted":true},"execution_count":167,"outputs":[]},{"cell_type":"code","source":"median(eukaryote$\"GC%\")","metadata":{"_uuid":"863ebde74c52d3e1b7d1ee51eca29f97004a557d","_cell_guid":"215becea-a051-40d6-9ff9-d3173818e9f2","execution":{"iopub.status.busy":"2022-05-19T03:37:04.259536Z","iopub.execute_input":"2022-05-19T03:37:04.262095Z","iopub.status.idle":"2022-05-19T03:37:04.276790Z"},"trusted":true},"execution_count":168,"outputs":[]},{"cell_type":"code","source":"median(prokaryote$\"GC%\")","metadata":{"_uuid":"283565f4be167094ae90812a01c9d66f4016473a","_cell_guid":"38846277-2402-44b7-917d-5b605570f952","execution":{"iopub.status.busy":"2022-05-19T03:37:04.279575Z","iopub.execute_input":"2022-05-19T03:37:04.282114Z","iopub.status.idle":"2022-05-19T03:37:04.296746Z"},"trusted":true},"execution_count":169,"outputs":[]},{"cell_type":"code","source":"median(viruses$\"GC%\")","metadata":{"_uuid":"dceb4ac722088fc10011e0cc5f61fd2220628ab1","_cell_guid":"04835b3a-4eaa-4efd-ab5e-df7ebfe9e69a","execution":{"iopub.status.busy":"2022-05-19T03:37:04.298804Z","iopub.execute_input":"2022-05-19T03:37:04.302066Z","iopub.status.idle":"2022-05-19T03:37:04.326513Z"},"trusted":true},"execution_count":170,"outputs":[]},{"cell_type":"markdown","source":"## Comparing trends in the genomes of Viruses, Prokaryotes and Eukaryotes\n\nHere I put all three organism types into a single dataframe for comparing the GC content and genome size (mb) across the domains.","metadata":{"_uuid":"3fce1560d8417542e8a21b5143f8e5212a7a2e2e","_cell_guid":"770eda53-f599-440b-b7ae-735a57a8afa8"}},{"cell_type":"code","source":"all_genomes = data.frame(domain= \"virus\", \n\t\t\t\t\t\t\tGC = viruses$\"GC%\", \n\t\t\t\t\t\t\tsize_mb = viruses$\"Size(Mb)\" )\n\nall_genomes = rbind(all_genomes, data.frame(domain=\"prokaryote\", \n\t\t\t\t\t\t\t\t\t\t\t\tGC = prokaryote$\"GC%\", \n\t\t\t\t\t\t\t\t\t\t\t\tsize_mb = prokaryote$\"Size(Mb)\"))\nall_genomes = rbind(all_genomes, data.frame(domain=\"eukaryote\", \n\t\t\t\t\t\t\t\t\t\t\t\tGC = eukaryote$\"GC%\", \n\t\t\t\t\t\t\t\t\t\t\t\tsize_mb = eukaryote$\"Size(Mb)\"))","metadata":{"_uuid":"ac36fac123ce3e268f6f11aa9b8dd2af5470fdb9","_cell_guid":"9df01aaa-a21c-407d-a8a1-69eda9b5ad15","execution":{"iopub.status.busy":"2022-05-19T03:37:04.329305Z","iopub.execute_input":"2022-05-19T03:37:04.331970Z","iopub.status.idle":"2022-05-19T03:37:04.351858Z"},"trusted":true},"execution_count":171,"outputs":[]},{"cell_type":"markdown","source":"First a simple boxplot of the gc content across the categories","metadata":{"_uuid":"543be8a8a2e027a9af52d9c1e31861adaa6d7a31","_cell_guid":"cc0b61c8-6a92-4a94-a7d4-7767d2283431"}},{"cell_type":"code","source":"boxplot(GC ~ domain , data = all_genomes,\n            main = \"%GC content of sequenced Virus, Prokaryote and Eukaryote genomes\")","metadata":{"_uuid":"983c7f277f292e5db62d2865a93264183ce17d96","_cell_guid":"f7525d8f-f692-4bdf-be77-9b4542ab1d82","execution":{"iopub.status.busy":"2022-05-19T03:37:04.354815Z","iopub.execute_input":"2022-05-19T03:37:04.357679Z","iopub.status.idle":"2022-05-19T03:37:04.448015Z"},"trusted":true},"execution_count":172,"outputs":[]},{"cell_type":"markdown","source":"I find that by overlaying a scatter plot on top of the boxplot, we can visualize the distribution of the data more effectively. Below I accomplish this using ggplot to produce a boxplt via `geom_boxplot` and then I overlay translucent, jittered (moved to not overlap to heavily) dots via `geom_point`. The rest of the lines are visual blandishment.","metadata":{"_uuid":"819f081ef4c7cefd2198faa3b1b2f1ee1a7539fd","_cell_guid":"69094e76-2c25-4698-880f-e467c645645e"}},{"cell_type":"code","source":"#plot of the gc content by domain\nggplot(all_genomes, aes(x=domain, y=GC, fill=domain, color=domain))+\n\tgeom_boxplot(outlier.shape = NA) +\n\tgeom_point(shape = 21, alpha = 0.1, position=position_jitterdodge())+\n\tscale_fill_manual(values = rep(NA,3)) +\n\ttheme_fivethirtyeight() +\n\tscale_colour_economist() +\n    labs(title = \"%GC content of sequenced Virus, Prokaryote and Eukaryote genomes\", \n            y = \"%GC content\",\n            x = \"domain\")","metadata":{"_uuid":"f91c506a2b8ec9c5fce1026ebe03dd01e4ed04ba","_cell_guid":"caf9a337-70c3-4c77-8f9e-63ffc8931a89","execution":{"iopub.status.busy":"2022-05-19T03:37:04.451583Z","iopub.execute_input":"2022-05-19T03:37:04.454140Z","iopub.status.idle":"2022-05-19T03:37:06.322438Z"},"trusted":true},"execution_count":173,"outputs":[]},{"cell_type":"markdown","source":"Here I run a simple analysis of variance to compare the mean %GC content across the domains.\n\nH0 = The mean %GC content is the same across the domains\nHa = The mean %GC content differs across the domains","metadata":{"_uuid":"48ebcbe5f5ce0b24e4fbc9fc9d7b5fa49c319312","_cell_guid":"92961f7b-2354-4160-8a7a-ea4799d0e7a4"}},{"cell_type":"code","source":"gc_anova = aov(GC ~ domain , data=all_genomes)\nsummary(gc_anova)","metadata":{"_uuid":"a8cbb26e4883a8a861ce9be4cfffba13143d750c","_cell_guid":"415ffea2-463b-434e-a9c6-5ed15ae5136a","execution":{"iopub.status.busy":"2022-05-19T03:37:06.328672Z","iopub.execute_input":"2022-05-19T03:37:06.332099Z","iopub.status.idle":"2022-05-19T03:37:06.377165Z"},"trusted":true},"execution_count":174,"outputs":[]},{"cell_type":"markdown","source":"Therefore we can reject the null hypothesis.\n\nUsing a Tukey's honest significant difference test we can then assess which categories have significant differences from one another.","metadata":{"_uuid":"fd92848a44834d9bbd42e71925c9867068b58196","_cell_guid":"cd9f9892-7018-4f0e-b85d-c8d32771a714"}},{"cell_type":"code","source":"TukeyHSD(gc_anova)","metadata":{"_uuid":"c2e8fa25958e604ad5a96b91968192749a65acd5","_cell_guid":"ccc0a902-1b22-4a60-97ee-bec8277bcccf","execution":{"iopub.status.busy":"2022-05-19T03:37:06.382412Z","iopub.execute_input":"2022-05-19T03:37:06.390485Z","iopub.status.idle":"2022-05-19T03:37:06.450129Z"},"trusted":true},"execution_count":175,"outputs":[]},{"cell_type":"markdown","source":"From those results we can conclude that there is no significant difference between the virus and eukaryote %GC content, while the prokaryote %GC content significantly differs from the other two groups.\n\nBelow I repeat the ANOVA process with an analysis of genome size across the three genome categorie.","metadata":{"_uuid":"32dddcac280a74362a12247d16d4baded58a4023","_cell_guid":"4dc5ff8a-552b-4a01-9e88-5fe7c9da4ca2"}},{"cell_type":"code","source":"#differences in size? should be obvious\nmb_anova = aov(size_mb ~ domain , data=all_genomes)\nsummary(mb_anova)","metadata":{"_uuid":"f1f7afab91a63dc55e09ced6c100126905e84751","_cell_guid":"9c3161ab-92d8-4402-b5d8-f89768bef25c","execution":{"iopub.status.busy":"2022-05-19T03:37:06.455167Z","iopub.execute_input":"2022-05-19T03:37:06.459395Z","iopub.status.idle":"2022-05-19T03:37:06.526087Z"},"trusted":true},"execution_count":176,"outputs":[]},{"cell_type":"markdown","source":"Again we can assess where the significant differences lie","metadata":{"_uuid":"38723215129e9394a396beaaaafb5417481d5f38","_cell_guid":"92043eb7-0d3a-4a2a-8e55-4cfc362e1dcd"}},{"cell_type":"code","source":"TukeyHSD(mb_anova)","metadata":{"_uuid":"2ab4f5ce0cf9e8066c8708bbdeb456ec10d8d4c8","_cell_guid":"8ccba261-a435-4e84-bafa-c47fbb83971b","execution":{"iopub.status.busy":"2022-05-19T03:37:06.530837Z","iopub.execute_input":"2022-05-19T03:37:06.534925Z","iopub.status.idle":"2022-05-19T03:37:06.621125Z"},"trusted":true},"execution_count":177,"outputs":[]},{"cell_type":"markdown","source":"The virus and prokaryote categories are on average not significantly different in terms of genome size, while the mean eukaryote genome size is significantly different from the other two groups.","metadata":{"_uuid":"01b08d4f9e3e4453e0e4e3dce7ba15c88c27bd47","_cell_guid":"69ff1bc8-ec4e-4584-ba3d-15115e107236"}},{"cell_type":"code","source":"ggplot(all_genomes, aes(x=size_mb, y=GC, shape=domain, colour=domain)) +\n        geom_point() +\n        scale_colour_brewer(palette=\"Set1\") +\n        coord_trans(x = \"log10\") +\n        labs(title = \"GC content vs genome size (mb) of genome sequences stored on NCBI\", \n        \t\ty = \"%GC content\", \n        \t\tx = \"Genome size (mb) - Note this axis is log scaled\") +\n        theme(axis.text.x = element_text(angle = 90, hjust = 1))","metadata":{"_uuid":"b7c19279dc6a92279b79dd5e8b7b7fb216e45c52","_cell_guid":"716e874e-b926-4887-939a-08911bb2da26","execution":{"iopub.status.busy":"2022-05-19T03:37:06.626106Z","iopub.execute_input":"2022-05-19T03:37:06.630313Z","iopub.status.idle":"2022-05-19T03:37:07.557166Z"},"trusted":true},"execution_count":178,"outputs":[]},{"cell_type":"markdown","source":"Here I have [a large chunk of code](https://gist.github.com/dgrtwo/eb7750e74997891d7c20#file-geom_flat_violin-r) that adds a new plot type into the libary of avaliable ggplot options. It lets us create a half violin plot.","metadata":{"_uuid":"a38b31e0441d077c4cf8e24290269b930415dbd6","_cell_guid":"e4071b6e-97ba-4af6-880b-e06201e2243d"}},{"cell_type":"code","source":"#######\n## load source code\n#locally:\n# source(\"geom_flat_violin.R\")\n\n## sourced from github \"dgrtwo/geom_flat_violin.R\n## https://gist.github.com/dgrtwo/eb7750e74997891d7c20#file-geom_flat_violin-r\n\n# I have pasted in the raw function below for the use in this kernel\n#######\n\n\n\"%||%\" <- function(a, b) {\n  if (!is.null(a)) a else b\n}\n\ngeom_flat_violin <- function(mapping = NULL, data = NULL, stat = \"ydensity\",\n                        position = \"dodge\", trim = TRUE, scale = \"area\",\n                        show.legend = NA, inherit.aes = TRUE, ...) {\n  layer(\n    data = data,\n    mapping = mapping,\n    stat = stat,\n    geom = GeomFlatViolin,\n    position = position,\n    show.legend = show.legend,\n    inherit.aes = inherit.aes,\n    params = list(\n      trim = trim,\n      scale = scale,\n      ...\n    )\n  )\n}\n\n#' @rdname ggplot2-ggproto\n#' @format NULL\n#' @usage NULL\n#' @export\nGeomFlatViolin <-\n  ggproto(\"GeomFlatViolin\", Geom,\n          setup_data = function(data, params) {\n            data$width <- data$width %||%\n              params$width %||% (resolution(data$x, FALSE) * 0.9)\n            \n            # ymin, ymax, xmin, and xmax define the bounding rectangle for each group\n            data %>%\n              group_by(group) %>%\n              mutate(ymin = min(y),\n                     ymax = max(y),\n                     xmin = x,\n                     xmax = x + width / 2)\n          },\n          \n          draw_group = function(data, panel_scales, coord) {\n            # Find the points for the line to go all the way around\n            data <- transform(data, xminv = x,\n                              xmaxv = x - violinwidth * (xmax - x)) # change x - violinwidth to + for violin to right\n            \n            # Make sure it's sorted properly to draw the outline\n            newdata <- rbind(plyr::arrange(transform(data, x = xminv), y),\n                             plyr::arrange(transform(data, x = xmaxv), -y))\n            \n            # Close the polygon: set first and last point the same\n            # Needed for coord_polar and such\n            newdata <- rbind(newdata, newdata[1,])\n            \n            ggplot2:::ggname(\"geom_flat_violin\", GeomPolygon$draw_panel(newdata, panel_scales, coord))\n          },\n          \n          draw_key = draw_key_polygon,\n          \n          default_aes = aes(weight = 1, colour = \"grey20\", fill = \"white\", size = 0.5,\n                            alpha = NA, linetype = \"solid\"),\n          \n          required_aes = c(\"x\", \"y\")\n)","metadata":{"_uuid":"ddaad9e6eeb9a36925e0e7d63e2040b5e68fe558","_cell_guid":"be1800b6-5be7-4b71-9f83-8953cfb3db34","execution":{"iopub.status.busy":"2022-05-19T03:37:07.559795Z","iopub.execute_input":"2022-05-19T03:37:07.562802Z","iopub.status.idle":"2022-05-19T03:37:07.565420Z"},"trusted":true},"execution_count":179,"outputs":[]},{"cell_type":"markdown","source":"## Violin and scatter plot of the %GC content \nBelow is a new plot style I have been experimenting with, utilizing the code we just ran above.  it is essentially half of a normal [violin plot](http://ggplot2.tidyverse.org/reference/geom_violin.html) and half a jittered scatter plot. I find that this does a good job of showing the distribution of the data via two parallel methods! Below I use it to plot the distribution of %GC content across the three groups of organisms. The violin shows the general distribution, while the scatter plot shows where the individual points fall.\n","metadata":{"_uuid":"099b391bd6a675f8edade952d49b012b3e30deac","_cell_guid":"854647a7-a1ba-4965-933f-8a2e2e132e73"}},{"cell_type":"code","source":"\n\nggplot(all_genomes, aes(x = domain, y = GC, fill=domain)) +\n\tgeom_flat_violin( colour=\"white\") +\n\tgeom_point(aes(x = as.numeric(domain) + .12, colour=domain) , \n\t\t\t\t\tsize = 0.01, \n\t\t\t\t\tpch=21,\n                    alpha = 0.1,\n\t\t\t\t\tposition = position_jitterdodge(jitter.width = .55, jitter.height = 0.01, dodge.width = 0.75)) +\n\ttheme_fivethirtyeight() +\n    labs(title = \"GC content of genome sequences stored on NCBI\", \n        \ty = \"%GC content\", \n        \tx = \"Genome size (mb)\")","metadata":{"_uuid":"cd1b0ccee0da49a607cfccb5bfc1ed328e08ea33","_cell_guid":"a2b3a018-647f-44ce-b40e-8f61674909ba","execution":{"iopub.status.busy":"2022-05-19T03:37:07.568541Z","iopub.execute_input":"2022-05-19T03:37:07.571106Z","iopub.status.idle":"2022-05-19T03:37:09.064668Z"},"trusted":true},"execution_count":180,"outputs":[]},{"cell_type":"markdown","source":"**#My Own Annalysis:**\n**My hypothesis:** Insects have higher  amount of DNA contained in a haploid genome than Land Plants\n\n","metadata":{}},{"cell_type":"code","source":"#Now I want Graph the Chromosome and Scaffold level only\neukaryote = raw_eukaryote[raw_eukaryote$Level == \"Chromosome\" || raw_eukaryote$Level == \"Scaffold\",]\nhead(eukaryote)\n","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.067120Z","iopub.execute_input":"2022-05-19T03:37:09.069155Z","iopub.status.idle":"2022-05-19T03:37:09.128696Z"},"trusted":true},"execution_count":181,"outputs":[]},{"cell_type":"markdown","source":"*Now I chose only Land Plants data to use*","metadata":{}},{"cell_type":"code","source":"eukaryote_LandPlant = eukaryote[eukaryote$\"Organism Groups\" == \"Eukaryota;Plants;Land Plants\",]\nhead(eukaryote_LandPlant)\ncount(eukaryote_LandPlant)","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.131110Z","iopub.execute_input":"2022-05-19T03:37:09.132864Z","iopub.status.idle":"2022-05-19T03:37:09.195612Z"},"trusted":true},"execution_count":182,"outputs":[]},{"cell_type":"markdown","source":"*I do not know the main size(Mb) of Land Plant, so I will graph a general bar before going to the final bar.*","metadata":{}},{"cell_type":"code","source":"eukaryote_LandPlant = eukaryote_LandPlant[eukaryote_LandPlant$\"Size(Mb)\" > 5,]\n","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.199286Z","iopub.execute_input":"2022-05-19T03:37:09.203030Z","iopub.status.idle":"2022-05-19T03:37:09.206541Z"},"trusted":true},"execution_count":183,"outputs":[]},{"cell_type":"code","source":"#Graph the raw Graph before going further\nqplot(eukaryote_LandPlant$\"Size(Mb)\") +\n    labs(title = \"Histogram of Land Plant genome Size(Mb) content in Scaffold Level\", x= \"Size(Mb) content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.209868Z","iopub.execute_input":"2022-05-19T03:37:09.213683Z","iopub.status.idle":"2022-05-19T03:37:09.660961Z"},"trusted":true},"execution_count":184,"outputs":[]},{"cell_type":"markdown","source":"**Because most of the data is less than 600**\n\nI chose the data less than 500 only","metadata":{}},{"cell_type":"code","source":"eukaryote_LandPlant = eukaryote_LandPlant[eukaryote_LandPlant$\"Size(Mb)\" > 5,]\neukaryote_LandPlant = eukaryote_LandPlant[eukaryote_LandPlant$\"Size(Mb)\" < 500,]","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.664822Z","iopub.execute_input":"2022-05-19T03:37:09.668700Z","iopub.status.idle":"2022-05-19T03:37:09.683057Z"},"trusted":true},"execution_count":185,"outputs":[]},{"cell_type":"code","source":"# This is the final graph for Plant data:\nqplot(eukaryote_LandPlant$\"Size(Mb)\") +\n    labs(title = \"Histogram of Land Plant genome Size(Mb) content in Scaffold Level\", x= \"Size(Mb) content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.686955Z","iopub.execute_input":"2022-05-19T03:37:09.690952Z","iopub.status.idle":"2022-05-19T03:37:09.987346Z"},"trusted":true},"execution_count":186,"outputs":[]},{"cell_type":"markdown","source":"**Read the Insect data**:","metadata":{}},{"cell_type":"code","source":"eukaryote_Insects = eukaryote[eukaryote$\"Organism Groups\" == \"Eukaryota;Animals;Insects\",]\nhead(eukaryote_Insects)\ncount(eukaryote_Insects)","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:09.991100Z","iopub.execute_input":"2022-05-19T03:37:09.994775Z","iopub.status.idle":"2022-05-19T03:37:10.059580Z"},"trusted":true},"execution_count":187,"outputs":[]},{"cell_type":"code","source":"#\neukaryote_Insects = eukaryote_Insects[eukaryote_Insects$\"Size(Mb)\" > 5,]","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.062761Z","iopub.execute_input":"2022-05-19T03:37:10.065599Z","iopub.status.idle":"2022-05-19T03:37:10.074153Z"},"trusted":true},"execution_count":188,"outputs":[]},{"cell_type":"code","source":"#Graph the raw Graph before going further\nqplot(eukaryote_Insects$\"Size(Mb)\") +\n    labs(title = \"Histogram of Land Plant genome Size(Mb) content in Scaffold Level\", x= \"Size(Mb) content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.077655Z","iopub.execute_input":"2022-05-19T03:37:10.080467Z","iopub.status.idle":"2022-05-19T03:37:10.360618Z"},"trusted":true},"execution_count":189,"outputs":[]},{"cell_type":"code","source":"#Let clean the data before runing it. the size will be higher than 5 to discard the random or data with 0\neukaryote_Insects = eukaryote_Insects[eukaryote_Insects$\"Size(Mb)\" > 5,]\neukaryote_Insects = eukaryote_Insects[eukaryote_Insects$\"Size(Mb)\" < 500,]","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.363053Z","iopub.execute_input":"2022-05-19T03:37:10.365652Z","iopub.status.idle":"2022-05-19T03:37:10.385598Z"},"trusted":true},"execution_count":190,"outputs":[]},{"cell_type":"code","source":"#Graph the raw Graph before going further\nqplot(eukaryote_Insects$\"Size(Mb)\") +\n    labs(title = \"Histogram of Land Plant genome Size(Mb) content in Scaffold Level\", x= \"Size(Mb) content\", y = \"Frequency\") +\n\ttheme_fivethirtyeight()","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.388168Z","iopub.execute_input":"2022-05-19T03:37:10.390762Z","iopub.status.idle":"2022-05-19T03:37:10.658715Z"},"trusted":true},"execution_count":191,"outputs":[]},{"cell_type":"markdown","source":"Combine data of Land Plant and the Insect","metadata":{}},{"cell_type":"code","source":"both_genomes = data.frame(domain= \"eukaryote_Insects\",\n\t\t\t\t\t\t\t\t\t\t\t\tsize_mb = eukaryote_Insects$\"Size(Mb)\")\nboth_genomes = rbind(both_genomes, data.frame(domain=\"eukaryote_LandPlant\", \n\t\t\t\t\t\t\t\t\t\t\t\tsize_mb = eukaryote_LandPlant$\"Size(Mb)\"))","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.661343Z","iopub.execute_input":"2022-05-19T03:37:10.662994Z","iopub.status.idle":"2022-05-19T03:37:10.684665Z"},"trusted":true},"execution_count":192,"outputs":[]},{"cell_type":"code","source":"#This is how the new data look like:\nboth_genomes","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.687317Z","iopub.execute_input":"2022-05-19T03:37:10.689822Z","iopub.status.idle":"2022-05-19T03:37:10.741274Z"},"trusted":true},"execution_count":193,"outputs":[]},{"cell_type":"code","source":"boxplot(size_mb ~ domain , data = both_genomes,\n            main = \"Size(Mb) of Both Land Plant and Insect \")","metadata":{"execution":{"iopub.status.busy":"2022-05-19T03:37:10.743876Z","iopub.execute_input":"2022-05-19T03:37:10.746442Z","iopub.status.idle":"2022-05-19T03:37:10.826081Z"},"trusted":true},"execution_count":194,"outputs":[]},{"cell_type":"markdown","source":"According to the graph, The hypothesis that Insects contain higher amount of DNA than Land Plant is rejected","metadata":{}}]}